Use new internal modification timestamp as part of the hash key for scales.

maethu opened the pull request:
Depends on #150
Fixes #151 

This PR changes the "date time" factor for the hash_key generation via the @@images view.

Instead of using `_p_mtime` it uses the new introduces `modified` attribute. 
This gives us greater control over the generated hash_key and how to invalidate it.

For example, since the catalog does the indexing async in a precommit hook, the `_p_mtime` was basically not yet there, or too old. So the catalog always stored the prev. hash_key and not the one from the current image. 

Potential Issue:
Since this change introduces a new hash_key for all images, event if not changed, in theory a lot of image scales have to be generated. 

❓ ~~Question~~❓ : 
~~Either the namefile/namedimage or the @@images view could handle the case, if there is no modified timestamp available and fallback to `_p_mtime`. This way only new and modified images would use the new logic and old ones do not need to touched/migrated.~~

maethu commented:
@jenkins-plone-org please run jobs

mauritsvanrees commented:
> Potential Issue: Since this change introduces a new hash_key for all images, event if not changed, in theory a lot of image scales have to be generated.
> 
> ❓ Question❓ : Either the namefile/namedimage or the @@images view could handle the case, if there is no modified timestamp available and fallback to `_p_mtime`. This way only new and modified images would use the new logic and old ones do not need to touched/migrated.

I think it would fail for older namedfiles/images, as they have no `modified` attribute.  But maybe they inherit it from higher up.

How about storing the timestamp on `_modified` and then have an extra property?

```
@property
def modified(self):
    # We may need to use base_hasattr here instead, if you see that Zope
    # tries to inherit the property from objects higher in the acquisition chain.
    # I expect so.
    if hasattr(self, "_modified):
        return self._modified
    # Fall back to modification time in database.
    return self._p_mtime
```

I guess this would also solve the potential issue you mention.

mauritsvanrees commented on plone/namedfile/tests/test_scaling.py:
```diff
@@ -650,11 +649,19 @@ def testScaleInvalidation(self):
+
+        # Test changing the fields modified timestamp invalidates scales
+        self.item.image.modified = (dt + 1).millis()
+        scale1b = self.scaling.scale("image", width=100, height=80)
+        scale2b = self.scaling.scale("image", width=80, height=60)
```
Can you rename these to `scale1c` and `scale2c`?  Now you are overriding the variables  from a few lines up, which may be confusing.
Or use other, longer names.

maethu commented:
> > Potential Issue: Since this change introduces a new hash_key for all images, event if not changed, in theory a lot of image scales have to be generated.
> > ❓ Question❓ : Either the namefile/namedimage or the @@images view could handle the case, if there is no modified timestamp available and fallback to `_p_mtime`. This way only new and modified images would use the new logic and old ones do not need to touched/migrated.
> 
> I think it would fail for older namedfiles/images, as they have no `modified` attribute. But maybe they inherit it from higher up.
> 
> How about storing the timestamp on `_modified` and then have an extra property?
> 
> ```
> @property
> def modified(self):
>     # We may need to use base_hasattr here instead, if you see that Zope
>     # tries to inherit the property from objects higher in the acquisition chain.
>     # I expect so.
>     if hasattr(self, "_modified):
>         return self._modified
>     # Fall back to modification time in database.
>     return self._p_mtime
> ```
> 
> I guess this would also solve the potential issue you mention.

Sounds good!
I implemented this [here](https://github.com/plone/plone.namedfile/pull/150/files)

maethu commented:
> # We may need to use base_hasattr here instead, if you see that Zope
> # tries to inherit the property from objects higher in the acquisition chain.
> # I expect so.

On my vanilla Plone the NameFile instances are not AQ wrapped anywhere. If there is a change it might be, I can use base_hasattr

mauritsvanrees reviewed (approved):
LGTM.

One minor comment is that I would not call this and the already merged PR internal changes.  Technically this may be true, but since this changes the way data is stored, I would call it a feature release.
You could even argue that this is a (potentially) breaking change, that we should not include in 6.0.  But I think it is okay, and I will include it in soon-to-be-released 6.0.7.
I hope I won't regret this. ;-)  And I may be more strict in the future.

I will rename the news snippets after merge.
